Add ApplicationInsights Migration Guide to Azure.Monitor.Query (#27235)
Azure/azure-sdk-for-net: sdk/monitor/Azure.Monitor.Query/MigrationGuideApplicationInsights.md

#### Consistency

There are now methods with similar names, signatures, and locations to create senders and receivers. The result is consistency and predictability on the various features of the library.

### Client constructors and authentication

In `Microsoft.Azure.ApplicationInsights.Query` v1.0.0:

```csharp
using Microsoft.Azure.ApplicationInsights.Query;
using Microsoft.Rest;
using Microsoft.Rest.Azure.Authentication;

// code omitted for brevity

var adSettings = new ActiveDirectoryServiceSettings
{
    AuthenticationEndpoint = new Uri("https://login.microsoftonline.com"),
    TokenAudience = new Uri("https://api.loganalytics.io/"),
    ValidateAuthority = true
};

ServiceClientCredentials credentials = ApplicationTokenProvider.LoginSilentAsync(
    "<domainId or tenantId>", "<clientId>", "<clientSecret>", adSettings).GetAwaiter().GetResult();
var client = new ApplicationInsightsDataClient(credentials);
```

In `Azure.Monitor.Query` v1.0.x:

```csharp

> using Azure.Monitor.Query;
> using Azure.Identity;

Reviewer: ```suggestion
using Azure.Identity;
using Azure.Monitor.Query;
```

// code omitted for brevity
QueryResults results = await client.Query.ExecuteAsync("<appId>", "AzureActivity | top 10 by TimeGenerated");
```

In `Azure.Monitor.Query` v1.0.x:

- The `QueryBody` is flattened. Users are expected to pass the Kusto query directly to the API.
- The `timespan` parameter is now required, which helps to avoid querying over the entire data set.

```csharp
using Azure;
using Azure.Monitor.Query;
using Azure.Monitor.Query.Models;

// code omitted for brevity

Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(
	workspaceId,
	"AzureActivity | top 10 by TimeGenerated",
	new QueryTimeRange(TimeSpan.FromDays(1)));
```

### A note about the Events API

The `Microsoft.Azure.ApplicationInsights.Query` package includes an Events API, which is just a different "API head" on the same logs data. It enables the querying of some logs in Application Insights without writing Kusto queries. The API translates to Kusto queries in the background. The same data can be accessed via regular Kusto queries, as shown in the preceding example.

### Query metrics from a resource

In `Microsoft.Azure.ApplicationInsights.Query` v1.0.0:

```csharp

> using Microsoft.Azure.ApplicationInsights.Query.Models
> 
> MetricsResult result = client.Metrics.GetAsync("<appId>", "<metricId>").Result;

Reviewer: ```suggestion
using Microsoft.Azure.ApplicationInsights.Query.Models;

// code omitted for brevity

MetricsResult result = client.Metrics.GetAsync("<appId>", "<metricId>").Result;
```

```

In `Azure.Monitor.Query` v1.0.x:

```csharp

> using Azure;

Reviewer: ```suggestion
```

> var client = new MetricsQueryClient(credential);

Reviewer: This client belongs in the Azure.Monitor.Query sample below.

> using Azure.Monitor.Query;

Reviewer: ```suggestion
using Azure.Identity;
using Azure.Monitor.Query;
```

using Azure.Monitor.Query.Models;

// code omitted for brevity

var credential = new ClientSecretCredential("<domainId or tenantId>", "<clientId>", "<clientSecret>");
var client = new MetricsQueryClient(credential);

> MetricsQueryResult result = await client.QueryResourceAsync("<resourceId>",new[]{ "<metricName>" }).Value;

Reviewer: ```suggestion
MetricsQueryResult result = await client.QueryResourceAsync("<resourceId>", new[]{ "<metricName>" }).Value;
```

#### Consistency

There are now methods with similar names, signatures, and locations to create senders and receivers. The result is consistency and predictability on the various features of the library.

### Client constructors and authentication

In `Microsoft.Azure.ApplicationInsights.Query` v1.0.0:

```csharp
using Microsoft.Azure.ApplicationInsights.Query;
using Microsoft.Rest;
using Microsoft.Rest.Azure.Authentication;

// code omitted for brevity

var adSettings = new ActiveDirectoryServiceSettings
{
    AuthenticationEndpoint = new Uri("https://login.microsoftonline.com"),
    TokenAudience = new Uri("https://api.loganalytics.io/"),
    ValidateAuthority = true
};

ServiceClientCredentials credentials = ApplicationTokenProvider.LoginSilentAsync(
    "<domainId or tenantId>", "<clientId>", "<clientSecret>", adSettings).GetAwaiter().GetResult();
var client = new ApplicationInsightsDataClient(credentials);

> ```

Reviewer: ```suggestion
```

```

In `Azure.Monitor.Query` v1.0.x:

```csharp
using Azure.Monitor.Query;
using Azure.Identity;

// code omitted for brevity

> var credential = new ClientSecretCredential("<domainId or tenantId>", "<clientId>", "<clientSecret>");

Reviewer: ```suggestion

var credential = new ClientSecretCredential("<domainId or tenantId>", "<clientId>", "<clientSecret>");
```

```

In `Azure.Monitor.Query` v1.0.x:

- The `QueryBody` is flattened. Users are expected to pass the Kusto query directly to the API.
- The `timespan` attribute is now required, which helped to avoid querying over the entire data set.

```csharp
using Azure;
using Azure.Monitor.Query;
using Azure.Monitor.Query.Models;

// code omitted for brevity

Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(
	workspaceId,
	"AzureActivity | top 10 by TimeGenerated",
	new QueryTimeRange(TimeSpan.FromDays(1)));
```

### A note about the Events API

The `Microsoft.Azure.ApplicationInsights.Query` package includes an Events API, which is just a different "API head" on the same logs data. It enables the querying of some logs in Application Insights without writing Kusto queries. The API translates to Kusto queries in the background. The same data can be accessed via regular Kusto queries, as shown in the preceding example.

### Query metrics from a resource

In `Microsoft.Azure.ApplicationInsights.Query` v1.0.0:

```csharp
using Microsoft.Azure.ApplicationInsights.Query.Models

// code omitted for brevity

> HttpResponseMessage response = await client.QueryAsync("<appId>", "AzureActivity | top 10 by TimeGenerated").ConfigureAwait(false).Response;

Reviewer: Wrap this line to avoid lots of horizontal scrolling.

```

In `Azure.Monitor.Query` v1.0.x:

```csharp
using Azure;
using Azure.Monitor.Query;
using Azure.Monitor.Query.Models;

// code omitted for brevity

Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(
	workspaceId,
	"AzureActivity | top 10 by TimeGenerated",
	new QueryTimeRange(TimeSpan.FromDays(1)));
LogsTable table = response.Value.Table;
IReadOnlyList<LogsTableRow> rows = table.Rows;

foreach (LogsTableRow row in rows)
{
    Console.WriteLine(row.GetString(0)); // Access a particular element with index
    Console.WriteLine(row.GetTimeSpan(1)); 
    Console.WriteLine(row.ToString());
}

> ```

Reviewer: ```suggestion
```

var adSettings = new ActiveDirectoryServiceSettings
{
    AuthenticationEndpoint = new Uri("https://login.microsoftonline.com"),
    TokenAudience = new Uri("https://api.loganalytics.io/"),
    ValidateAuthority = true
};

ServiceClientCredentials credentials = ApplicationTokenProvider.LoginSilentAsync(
    "<domainId or tenantId>", "<clientId>", "<clientSecret>", adSettings).GetAwaiter().GetResult();
var client = new ApplicationInsightsDataClient(credentials);
```

In `Azure.Monitor.Query` v1.0.x:

```csharp
using Azure.Monitor.Query;
using Azure.Identity;

// code omitted for brevity

var credential = new ClientSecretCredential("<domainId or tenantId>", "<clientId>", "<clientSecret>");
var client = new LogsQueryClient(credential);
```

### Send a single query request

In `Microsoft.Azure.ApplicationInsights.Query` v1.0.0:

```csharp
using Microsoft.Azure.ApplicationInsights.Query.Models

// code omitted for brevity
QueryResults results = await client.Query.ExecuteAsync("<appId>", "AzureActivity | top 10 by TimeGenerated");
```

In `Azure.Monitor.Query` v1.0.x:

- The `QueryBody` is flattened. Users are expected to pass the Kusto query directly to the API.

> - The `timespan` attribute is now required, which helped to avoid querying over the entire data set.

Reviewer: ```suggestion
- The `timespan` parameter is now required, which helps to avoid querying over the entire data set.
```

```csharp
using Azure;
using Azure.Monitor.Query;
using Azure.Monitor.Query.Models;

// code omitted for brevity

Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(
	workspaceId,
	"AzureActivity | top 10 by TimeGenerated",
	new QueryTimeRange(TimeSpan.FromDays(1)));
```

### A note about the Events API

The `Microsoft.Azure.ApplicationInsights.Query` package includes an Events API, which is just a different "API head" on the same logs data. It enables the querying of some logs in Application Insights without writing Kusto queries. The API translates to Kusto queries in the background. The same data can be accessed via regular Kusto queries, as shown in the preceding example.

### Query metrics from a resource

In `Microsoft.Azure.ApplicationInsights.Query` v1.0.0:

```csharp

> using Microsoft.Azure.ApplicationInsights.Query.Models

Reviewer: This example looks very different from what the Python migration guide shows in this section: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/monitor/azure-monitor-query/migration_guide_app_insights.md#query-metrics-from-a-resource.

Shouldn't this call `client.Metrics.GetAsync(...)` instead of `client.QueryAsync(...)`?

// code omitted for brevity
HttpResponseMessage response = await client.QueryAsync(
  "<appId>", 
  "AzureActivity | top 10 by TimeGenerated").
  ConfigureAwait(false).Response;
```

In `Azure.Monitor.Query` v1.0.x:

```csharp
using Azure;
using Azure.Monitor.Query;
using Azure.Monitor.Query.Models;

// code omitted for brevity

> Response<LogsQueryResult> response = await client.QueryWorkspaceAsync(

Reviewer: For querying metrics, you'd obviously use a `MetricsQueryClient` instance here. But that part isn't clear.
